File: gaia-sdk-javascript/src/graphql/GaiaClient.ts
import {ClientOptions} from '../api/ClientOptions';
import {ITransporter} from '../api/ITransporter';
import Query from "./request/type/Query";
import Mutation from "./request/type/Mutation";
import Subscription from "./request/type/Subscription";
import {QueryResponse} from './GaiaResponse';
import {MutationResponse} from './GaiaResponse';
import {SubscriptionResponse} from './GaiaResponse';
import VariableRegistry from "../api/VariableRegistry";

export class GaiaClient {
    private readonly options: ClientOptions;
    private readonly transporter: ITransporter;

    constructor(options: ClientOptions, transporter: ITransporter) {
        this.options = options;
        this.transporter = transporter;
    }

    public queryNative(statement: string, variables: Record<string, any> = {}):Promise<QueryResponse> {
        const body = {statement, variables};
        return this.transporter.transport(this.options, body);
    }

    public query(request: Query): Promise<QueryResponse> {
        const [statement, variables] = this.getStatement("query", request);
        return this.queryNative(statement, variables);
    }

    public mutationNative(statement: string, variables: Record<string, any> = {}): Promise<MutationResponse> {
        const body = {statement, variables};
        return this.transporter.transport(this.options, body);
    }

    public mutation(request: Mutation): Promise<MutationResponse> {
        const [statement, variables] = this.getStatement("mutation", request);
        return this.mutationNative(statement, variables);
    }

    public subscriptionNative(statement: string, variables: Record<string, any> = {}):Promise<SubscriptionResponse> {
        const body = {statement, variables};
        return this.transporter.transport(this.options, body);
    }

    public subscription(request: Subscription): Promise<SubscriptionResponse> {
        const [statement, variables] = this.getStatement("subscription", request);
        return this.subscriptionNative(statement, variables);
    }


    private getStatement(name: string, type: Array<(_:VariableRegistry) => string>):[string, {}] {
        const registry = new VariableRegistry();
        const fields = type.map(e => e(registry)).join(" ");

        if (registry.getDatatypes().length == 0) {
            const statement = `${name} gaia { ${fields} }`;
            return [statement, registry.getVariables()];
        }
        const statement = `${name} gaia(${registry.getDatatypes().join(", ")}) { ${fields} }`;
        return [statement, registry.getVariables()];
    }

}

export type Struct = Record<string, any>
export type Long = String
export type Timestamp = String
export type Uuid = String

File: gaia-sdk-python/gaia_sdk/graphql/GaiaClient.py
from gaia_sdk.api.VariableRegistry import VariableRegistry
import itertools
import random
from typing import List, Callable

from gaia_sdk.api.transporter.abstract_transporter import ITransporter
from gaia_sdk.graphql.GaiaResponse import QueryResponse
from gaia_sdk.graphql.request.type import Query
from gaia_sdk.graphql.GaiaResponse import MutationResponse
from gaia_sdk.graphql.request.type import Mutation
from gaia_sdk.api.client_options import ClientOptions
from gaia_sdk.http.request.Payload import Payload


class GaiaClient(object):

    transporter: ITransporter
    options: ClientOptions

    def __init__(self, transporter: ITransporter, options: ClientOptions):
        self.transporter = transporter
        self.options = options

    def __eq__(self, other):
        return self.options == other.options

    def __repr__(self):
        return {'options': self.options}

    def execute_native(self, statement: str, variables: dict):
        payload = {"statement": statement, "variables": variables}
        return self.transporter.transport(self.options, Payload.json(payload)).json()

    def query(self, request: Query) -> QueryResponse:
        statement, variables = self.get_statement("query", request)
        return self.execute_native(statement, variables)

    def mutation(self, request: Mutation) -> MutationResponse:
        statement, variables = self.get_statement("mutation", request)
        return self.execute_native(statement, variables)

    def get_statement(self, name: str, values: List[Callable[[VariableRegistry], str]]):
        registry = VariableRegistry()
        fields = " ".join(map(lambda x: x(registry), values))
        if len(registry.getDatatypes()) == 0:
            statement = f'{name} gaia ' + '{ ' + fields + ' }'
            return statement, registry.getVariables()

        statement = name + ' gaia(' + ", ".join(registry.getDatatypes()) + ')  {' + fields + ' }'
        return statement, registry.getVariables()
